### PR TITLE
feat: optimistic transaction protocol

### LINKED ISSUE #593: Protocol compliant commits for Delta operations
# Description

When committing transactions to the Delta log, make sure we never end up in an inconsistent or corrupt state.

**Use Case**

When performing mutating operations against a Delta table we have to make sure we never corrupt the table / data.

### PR DESCRIPTION
**Pull request by roeap**
# Description

This PR adds a `ConflictChecker` struct for conflict resolution in cases of concurrent commit failures. The implementation is heavily inspired by the [reference implementation](https://github.com/delta-io/delta/blob/fe36a53f3c70c5f9c9b5052c12cd1703f495da97/core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala). So far we have most tests from spark that specifically target conflict resolution covered.

Working on this I thought a bit about what we may consider going forward, as we move through the protocol versions :).  In the end we could end up with three main structs that are involved in validating a commit.

* The existing `DataChecker`, which validates and potentially mutates data when writing data files to disk. (Currently supports invariants)
* The upcoming `ConflictChecker`, which checks if a commit can be re-tried in case of commit conflicts.
* A new `CommitChecker`, which does a-priory validation of the commit itself (e.g. append only and other rules covered by tests in  [spark](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala))

My hope is to get this PR merged right after we release `0.8.0`, so there is some time to fill some holes and fully leverage the new feature for `0.9.0`. 

If folks agree, I would open some issues and start work on some follow-ups..

## Follow-ups
* Extend `ConflictChecker` support conflict resolution for streaming transactions
* Implement `CommitChecker`
* Deprecate old commit function.
* Extend `DataChecker`. 
* Consolidate record batch writer implementations.

# Related Issue(s)

part of #593 

# Documentation

### CODE REVIEW

**Comment by houqp** [comment]
Thank you @roeap for picking up this work! I will take a closer look at it this weekend. :+1: from me for including datafusion-expr if needed.

**Comment by houqp** [comment]
The structure looks good to me :+1: When we introduce a proper expression handling abstraction, we should upgrade what we have in partition handling with it as well: https://github.com/delta-io/delta-rs/blob/cc62c49c905a0818e5d1edcd1383deb512dc2ac9/rust/src/partitions.rs#L12

File: rust/src/conflict_checker.rs
Comment (wjones127) [inline rust/src/conflict_checker.rs:339]: What would you think of keeping this behavior if datafusion feature isn't included? That way we could still have the option of not including datafusion. I think there are some use cases where we don't want to require it as a dependency.

Comment (roeap) [reply rust/src/conflict_checker.rs:339]: Absolutely - I was contemplating how to do this for a bit and parsing complex predicates would certainly benefit from some more advanced parsing. However right now we only have min max and nulls anyhow, which the current predicate parsing should be able to parse as well.

When we apply the filters my feeling is that we want to go the arrow route - i.e. put all files with bounds in  a record batch and use arrow kernels. I'll see how far I get using "just" arrow and as you said, falling back to the most conservative case is always an option.

File: rust/src/conflict_checker.rs
Comment (andrei-ionescu) [inline rust/src/conflict_checker.rs:13]: Please add a link to the official documentation on the Delta Lake possible conflicts.

I did find this one — https://docs.databricks.com/optimizations/isolation-level.html#write-conflicts-on-databricks — but I'm not sure how complete it is.

File: rust/src/conflict_checker.rs
Comment (andrei-ionescu) [inline rust/src/conflict_checker.rs:34]: I don't think `INSERT` / `INSERT` can result in this error. As far as I can understand from https://docs.databricks.com/optimizations/isolation-level.html#write-conflicts-on-databricks.

Maybe it will be useful to create that matrix from the given link and add it somewhere in this PR maybe on the enum explanation.

File: rust/src/conflict_checker.rs
Comment (andrei-ionescu) [inline rust/src/conflict_checker.rs:289]: Why forcing the `IsolationLevel::Serializable`? What about the other isolation levels?

File: rust/src/delta.rs
Comment (andrei-ionescu) [inline rust/src/delta.rs:1329]: Should there be some explanation added regarding `txn_id` in the docs above?

File: rust/src/delta.rs
Comment (andrei-ionescu) [inline rust/src/delta.rs:1361]: Changing this public method signature will result in backward compatibility issues. Any concerns? Is backward compatibility offered in the APIs?

File: rust/src/delta.rs
Comment (andrei-ionescu) [inline rust/src/delta.rs:1364]: In the master the default isolation level seems to be retrieved from table metadata: https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala#L1213

File: rust/src/delta.rs
Comment (andrei-ionescu) [inline rust/src/delta.rs:1415]: Ditto... This is a change that will break backward compatibility.

File: rust/src/delta.rs
Comment (andrei-ionescu) [inline rust/src/delta.rs:1482]: Is there a reason why the result is not used?

**Comment by roeap** [comment]
@wjones127 @houqp - things are finally taking shape around here :). As you can see this PR is quite sizeable already and while it also contains changes from #1154 it will stay large.

Thus a quick question as for your preference in reviewing this. I could leave it as is (just rebase once obove PR is merged). I could rearrange the commits, such that its easier to review, or i could dissect into multiple PRs. Happy to hear your thoughts!

I am also preparing a small design document that I will share soon.

**Comment by wjones127** [comment]
FYI I'm planning on reviewing this weekend. :)

File: rust/src/operations/filesystem_check.rs
Comment (roeap) [inline rust/src/operations/filesystem_check.rs:154]: I plan to do a pass through all our operations in the next Pr, where also I'll be focussing on testing the `commit` function, and addressing these kinds of TODOs.

File: rust/src/operations/transaction/state.rs
Comment (roeap) [inline rust/src/operations/transaction/state.rs:85]: Parsing predicate expressions is actually kind of broken. Regardless of the schema we pass in, int fields are always recognized as i64. I hope, that the abstractions in the core project will eventually help us out. Until then I plan to either do a follow-up in re-writing these expressions, or, if this is a bug, addressing this upstream in datafusion.

File: rust/src/operations/transaction/test_utils.rs
Comment (roeap) [inline rust/src/operations/transaction/test_utils.rs:1]: Yet another set of test utilities, and yet another TODO for follow up PR, to consolidate our test helpers...

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:736]: Does `read + no write` mean no data change? like an optimize?

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:812]:
```suggestion
        // current and concurrent transactions change the protocol version
```

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:91]:
```suggestion
    /// partition predicates by which files have been queried by the transaction
    ///
    /// If any new data files or removed data files match this predicate, the 
    /// transaction should fail.
    #[cfg(not(feature = "datafusion"))]
    read_predicates: Option<String>,
```

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:229]: Yeah, if there are multiple versions, should we just error for now? If there are multiple commits that beat our candidate, it seems like we should generate a WinningCommitSummary for each transaction after `read_version`, right? And then checking our candidate commit against each of them?

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:26]: ConcurrentAppend is what went wrong, not what you are trying to do, right?

```suggestion
    #[error("Commit failed: a concurrent transactions added new files.\nHelp: This transaction's query must be rerun to include the new data. Also, if you don't care to require this check to pass in the future, the isolation level can be set to Snapshot Isolation.")]
```

File: rust/src/operations/transaction/conflict_checker.rs
Comment (wjones127) [inline rust/src/operations/transaction/conflict_checker.rs:31]: I think similar changes could be applied to all the errors to make them more user-friendly.

```suggestion
    #[error("Commit failed: a concurrent transaction deleted a file this operation read.\nHelp: This transaction's query must be rerun to exclude the removed data. Also, if you don't care to require this check to pass in the future, the isolation level can be set to Snapshot Isolation.")]
```

File: rust/src/action/mod.rs
Comment (wjones127) [reply rust/src/action/mod.rs:666]: Or could you provide examples of which operations will?

**Review by wjones127 — CHANGES_REQUESTED** [review]
Overall looks like a great improvement. I think we want to be more conservative in cases where there are multiple concurrent transactions that beat the candidate one. Also the error messages could be improved.

File: rust/src/operations/transaction/conflict_checker.rs
Comment (roeap) [reply rust/src/operations/transaction/conflict_checker.rs]: Good point. Changed the logic, that out commit loop will always only check against the currently conflicting commit and as such the `WinningCommitSummary` will now summarize exactly one commit.

File: rust/src/action/mod.rs
Comment (roeap) [reply rust/src/action/mod.rs:659]: Agreed! If OK with you, I would defer exploring that to a follow up PR. To really get a feel for how that API should look like, I would like to work a bit more with operations that actually need the conflict resolution :).

Comment (wjones127) [reply rust/src/action/mod.rs:659]: That sounds good to me!

File: rust/src/action/mod.rs
Comment (roeap) [reply rust/src/action/mod.rs:666]: Our operations should actually never use this directly right now. I went through the spark code base, and it seems this is invoked mostly in cases when a plan querying multiple tables with spark has a delta table set as a sink.. This usually uses generic plans and does not invoke the delta operations directly, or rather when we are not too sure what actually happened.

Did not get too deep though. Personally I am a bit conflicted on how to proceed with this. on one hand it does nothing useful right now. On the other hand, we can keep it as a reminder to mindful of this. While right now I think datafusion does not yet have the concept of a sink in their plans, I believe they might be added fairly soon. At this point we may encounter situations, where a plan wants to write to delta, but we have no way of knowing what was scanned.

That said, I do hope that we will be able to inspect the plan metrics and figure out what was read anyhow. Especially since we report the scanned files in out scan operator.

**Comment by roeap** [comment]
@wjones127 - sorry for taking a long time to incorporate your feedback. I think it's mostly done now.

Mainly we now always check against the next version of commits. The questions around `read_whole_table` etc. I hope to address right after this.

As I plus I re-enabled an existing test for optimise, which we can now support again.

**Comment by wjones127** [comment]
I am realizing we have another transaction implementation called `DeltaTransaction`, which is what we are currently using in Python. This improved implementation doesn't seem to have a low-level API where the user can provide the added and removed files, which is what we need for the PyArrow-based writer.

So I think we need to consolidate those into `operations/transaction/mod.rs`, make them public, and switch the Python implementation. That sounds right to you?

**Review by wjones127 — APPROVED** [review]
This is looking good now.

The PySpark integration tests will be fixed soon. I should take a look at the sporadic failures in the Azure tests.

**Comment by roeap** [comment]
> So I think we need to consolidate those into `operations/transaction/mod.rs`, make them public, and switch the Python implementation. That sounds right to you?

Yes it does. So far I have deliberately kept the implementation of commits used in the operations module separate from the existing one on the table. Where I am not sure is if we need the `DeltaTransaction` to be a struct at all, I think this is somehow an artifact from the reference implementation, where the Traction is conceptually used kind of like a context in python (it's been a long time since I wrote scala :D). In the new trasaction module we just have the [`commit`](https://github.com/delta-io/delta-rs/blob/d9920aaf6c730fda2878c2271ed396a138243bb1/rust/src/operations/transaction.rs#L140-L156) function, which we should make public and which I think will give us what we need in python.

**Pull request merged**
